File: Lecture30/rwalk.cpp
#include "cptstd.hpp"
#include "linalg.hpp"
#include "random.hpp"
using namespace cpt;

Matrix<double,1> model(      // model function to fit a linear model
  const double step)
{
    Matrix<double,1> f(2);
    f[0] = 1;
    f[1] = step;
    return f;
}

int main() {
    cout << " Random walk in 1 dimension\n"
         << " --------------------------\n"
         << " Enter number of walkers: ";
    int n_walkers;
    cin >> n_walkers;
    cout << " Enter number of steps: ";
    int n_steps;
    cin >> n_steps;

    // Random number generator
    Random rng;
    cout << " Using " << rng.get_algorithm()
         << " with seed " << rng.get_seed() << endl;

    // vector of walker positions initialized to zero
    Matrix<double,1> x(n_walkers);

    // data vectors
    Matrix<double,1> steps(n_steps);     // to save step number i (time)
    Matrix<double,1> x2ave(n_steps);     // to accumulate x^2 values
    Matrix<double,1> sigma(n_steps);     // to accumulate fluctuations in x^2

    // Loop over walkers
    for (int walker = 0; walker < n_walkers; walker++) {

        // Loop over number of steps
        for (int step = 0; step < n_steps; step++) {

            // take a random step
            if (rng.rand() < 0.5)
                x[walker] += 1;
            else
                x[walker] -= 1;

            // accumulate data
            steps[step] = step + 1;
            double xx = x[walker] * x[walker];
            x2ave[step] += xx;
            sigma[step] += xx * xx;
        }
    }

    // average the squared displacements and their variances
    for (int step = 0; step < n_steps; step++) {
        x2ave[step] /= n_walkers;
        sigma[step] /= n_walkers;
    }
    for (int step = 0; step < n_steps; step++) {
        sigma[step] = sqrt(sigma[step] - x2ave[step] * x2ave[step]);
        if (sigma[step] == 0)   // happens for first step!
            sigma[step] = 1;
        if (n_walkers > 1)
            sigma[step] /= sqrt(double(n_walkers - 1));
    }

    // fit data to a straight line
    double a, b, sigma_a, sigma_b, chisqr;
    chi_square_fit(steps, x2ave, sigma, a, b, sigma_a, sigma_b, chisqr);

    cout << " Fit to a straight line <x^2> = a + b n\n"
         << " Intercept a = " << a << " +- " << sigma_a << '\n'
         << " Slope     b = " << b << " +- " << sigma_b << '\n'
         << " Chisqr/dof  = " << chisqr / (n_steps - 2.0) << endl;

    // store in file for plotting
    ofstream file("rwalk.data");
    for (int step = 0; step < n_steps; step++)
        file << steps[step] << '\t' << x2ave[step]
             << '\t' << sigma[step] << '\n';
    file.close();
    cout << " t, <x^2>, sigma in file rwalk.data" << endl;
}

File: Lecture30/rwalk.py
import math
import random
import cpt
import matplotlib.pyplot as plt

print " Random walk in 1 dimension"
print " --------------------------"
n_walkers = int(input(" Enter number of walkers: "))
n_steps = int(input(" Enter number of steps: "))

# walker positions initialized at x = 0
x = [ 0.0 ] * n_walkers

steps = [ 0.0 ] * n_steps       # to save step number i (time)
x2ave = [ 0.0 ] * n_steps       # to accumulate x^2 values
sigma = [ 0.0 ] * n_steps       # to accumulate fluctuations in x^2

# loop over walkers
for walker in range(n_walkers):

    # loop over number of steps
    for step in range(n_steps):

        # take a random step
        x[walker] += random.choice( (-1, 1) )

        # accumulate data
        steps[step] = step + 1.0
        x2ave[step] += x[walker]**2
        sigma[step] += x[walker]**4

# average the squared displacements and their variances
for step in range(n_steps):
    x2ave[step] /= float(n_walkers)
    sigma[step] /= float(n_walkers)
for step in range(n_steps):
    sigma[step] = math.sqrt(sigma[step] - x2ave[step]**2)
    if sigma[step] == 0.0:      # happens for first step!
        sigma[step] = 1.0
    if n_walkers > 1:
        sigma[step] /= math.sqrt(n_walkers - 1.0)

# fit data to a straight line
a, b, sigma_a, sigma_b, chisqr = cpt.chi_square_fit(steps, x2ave, sigma)
print " Fit to straight line <x^2> = a + b n"
print " Intercept a = " + repr(a) + " +- " + repr(sigma_a)
print " Slope     b = " + repr(b) + " +- " + repr(sigma_b)
print " Chisqr/dof  = " + repr(chisqr / (n_steps - 2.0))

# store in file for plotting
file = open("rwalk.data", "w")
for step in range(n_steps):
    file.write(repr(steps[step]) + "\t" + repr(x2ave[step]) +
                    "\t" + repr(sigma[step]) + "\n")
file.close()
print " t, <x^2>, sigma in file rwalk.data"

plt.scatter( steps, x2ave )

plt.show()
